pkg/controller/controller.go

```diff
@@ -0,0 +1,286 @@
+		func() {
+			defer r.queue.Done(key)
+			obj, exists, err := r.serviceInformer.GetIndexer().GetByKey(key)
+
+			if err != nil {
+				klog.Errorf("error retrieving service with key %q from the cache: %v", key, err)
+				// requeue the item to work on later
+				r.queue.AddRateLimited(key)
+				return
+			}
+			if !exists {
+				err := r.serviceDeleted(key)
+				if err != nil {
+					klog.Errorf("error deleting service %q in cluster %q: %v", key, r.clusterID, err)
+					r.queue.AddRateLimited(key)
+					return
+				}
+			} else {
+				err := r.serviceCreated(obj, key)
+				if err != nil {
+					klog.Errorf("error creating service %q in cluster %q: %v", key, r.clusterID, err)
+					r.queue.AddRateLimited(key)
+					return
+				}
+			}
+			r.queue.Forget(key)
+		}()
+	}
+}
+
+func (r *remoteCluster) serviceCreated(obj interface{}, key string) error {
+	klog.V(2).Infof("In remoteCluster serviceCreated for cluster %q, service %#v", r.clusterID, obj)
+	service := obj.(*core_v1.Service)
+	r.remoteServicesMutex.Lock()
+	defer r.remoteServicesMutex.Unlock()
+	if rs, exists := r.remoteServices[key]; !exists {
+		r.remoteServices[key] = r.createLightHouseCRD(service, r.clusterID)
+		klog.Infof("Creating a new MultiClusterService with service %q from cluster %q", key, r.clusterID)
+	} else {
+		klog.Infof("Adding service from cluster %q to existing MultiClusterService %q", r.clusterID, key)
+		rs.Spec.Items = append(rs.Spec.Items, r.createClusterServiceInfo(service, r.clusterID))
+	}
+	err := r.federator.Distribute(r.remoteServices[key])
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *remoteCluster) createLightHouseCRD(service *v1.Service, clusterID string) *mcservice.MultiClusterService {
+	return &mcservice.MultiClusterService{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      service.ObjectMeta.Name,
+			Namespace: service.ObjectMeta.Namespace,
+		},
+		Spec: mcservice.MultiClusterServiceSpec{
+			Items: []mcservice.ClusterServiceInfo{
+				r.createClusterServiceInfo(service, clusterID),
+			},
+		},
+	}
+}
+
+func (r *remoteCluster) createClusterServiceInfo(service *v1.Service, clusterID string) mcservice.ClusterServiceInfo {
+	return mcservice.ClusterServiceInfo{
+		ClusterID: clusterID,
+		ServiceIP: service.Spec.ClusterIP,
+	}
+}
+
+// serviceDeleted is called when an object is deleted
+func (r *remoteCluster) serviceDeleted(key string) error {
+	r.remoteServicesMutex.Lock()
+	defer r.remoteServicesMutex.Unlock()
+	if rs, exists := r.remoteServices[key]; exists {
+		if len(rs.Spec.Items) == 1 {
+			err := r.federator.Delete(r.remoteServices[key])
+			if err != nil {
+				return err
+			}
+			delete(r.remoteServices, key)
+			klog.Infof("Service %q has been deleted from cluster %q - deleted MultiClusterService", key, r.clusterID)
+		} else {
+			serviceList := rs.Spec.Items
+			for i, service := range serviceList {
+				if service.ClusterID == r.clusterID {
+					serviceList = append(serviceList[:i], serviceList[i+1:]...)
```

Comment:
You should break from the loop here.

Also using `append` will move elements over which can be inefficient. It's faster to swap the element to delete with the one at the end:

```
serviceList[i] = serviceList[len(serviceList)-1]
serviceList = serviceList[:len(serviceList)-1] 
```
This will not maintain the order but that shouldn't matter.